File: rust/x402/x402-core/src/subscription/facilitator.rs
//! `SubscriptionFacilitatorClient` — the subscription write/read endpoints the
//! Seller calls on the facilitator, with OK-ACCESS auth.
//!
//! Kept separate from the base [`crate::facilitator::FacilitatorClient`]
//! (verify/settle) so exact/upto impls need not implement subscription methods.
//! `OkxHttpFacilitatorClient` implements both.

use async_trait::async_trait;

use crate::error::X402Error;

use super::types::{
    ChangeResponse, ChangeSubscriptionRequest, ChargeResponse, ChargesResponse,
    CancelPendingChangeRequest, CancelSubscriptionRequest, CreateSubscriptionRequest,
    CreateSubscriptionResponse, PendingPlanChange, SubscriptionStatus, TxResultResponse,
};

/// Subscription endpoints (`/api/v6/pay/x402/subscriptions/*`). Writes are
/// authenticated with the Seller's OK-ACCESS API key; the server derives
/// `callerMerchantId` from the key (the SDK never sends merchantId).
#[async_trait]
pub trait SubscriptionFacilitatorClient: Send + Sync {
    /// `POST /subscriptions` — create and force-charge the first period.
    async fn create_subscription(
        &self,
        req: &CreateSubscriptionRequest,
    ) -> Result<CreateSubscriptionResponse, X402Error>;

    /// `POST /subscriptions/charge` — periodic charge.
    async fn charge(&self, sub_id: &str, sync_settle: bool)
        -> Result<ChargeResponse, X402Error>;

    /// `POST /subscriptions/change` — up/downgrade.
    async fn change_subscription(
        &self,
        req: &ChangeSubscriptionRequest,
    ) -> Result<ChangeResponse, X402Error>;

    /// `POST /subscriptions/cancel` — cancel with `CancelAuth`.
    async fn cancel_subscription(
        &self,
        req: &CancelSubscriptionRequest,
    ) -> Result<TxResultResponse, X402Error>;

    /// `POST /subscriptions/cancel-pending-change` — cancel a not-yet-effective
    /// downgrade (payer-signed).
    async fn cancel_pending_change(
        &self,
        req: &CancelPendingChangeRequest,
    ) -> Result<TxResultResponse, X402Error>;

    /// `POST /subscriptions/finalize-expired` — release the reservation of an
    /// ended subscription (any caller).
    async fn finalize_expired(&self, sub_id: &str) -> Result<TxResultResponse, X402Error>;

    /// `GET /subscriptions/detail?subId=` — current status.
    async fn get_subscription(&self, sub_id: &str) -> Result<SubscriptionStatus, X402Error>;

    /// `GET /subscriptions/charges?subId=&limit=&offset=` — charge ledger.
    async fn get_charges(
        &self,
        sub_id: &str,
        limit: u32,
        offset: u32,
    ) -> Result<ChargesResponse, X402Error>;

    /// `GET /subscriptions/pending?subId=` — the PENDING downgrade, or `None`.
    async fn get_pending_change(
        &self,
        sub_id: &str,
    ) -> Result<Option<PendingPlanChange>, X402Error>;
}

File: go/x402/subscription/facilitator.go
package subscription

import "context"

// SubscriptionFacilitatorClient is the seller's view of the nine OKX
// subscription endpoints. Write methods carry subId in the request body; read
// methods carry it (and paging) in the query string. It is deliberately
// separate from the base verify/settle facilitator client so one-shot schemes
// need not implement it.
type SubscriptionFacilitatorClient interface {
	// CreateSubscription creates a subscription and force-charges the first
	// period. POST /subscriptions.
	CreateSubscription(ctx context.Context, req *CreateSubscriptionRequest) (*CreateSubscriptionResponse, error)
	// Charge pulls the next periodic charge. POST /subscriptions/charge.
	Charge(ctx context.Context, req *ChargeRequest) (*ChargeResponse, error)
	// ChangeSubscription upgrades or downgrades a subscription.
	// POST /subscriptions/change.
	ChangeSubscription(ctx context.Context, req *ChangeSubscriptionRequest) (*ChangeResponse, error)
	// CancelSubscription relays a signed cancel. POST /subscriptions/cancel.
	CancelSubscription(ctx context.Context, req *CancelSubscriptionRequest) (*TxResultResponse, error)
	// CancelPendingChange relays a signed cancel of a pending downgrade.
	// POST /subscriptions/cancel-pending-change.
	CancelPendingChange(ctx context.Context, req *CancelPendingChangeRequest) (*TxResultResponse, error)
	// FinalizeExpired settles the expiry marker for a lapsed subscription.
	// POST /subscriptions/finalize-expired.
	FinalizeExpired(ctx context.Context, req *FinalizeExpiredRequest) (*TxResultResponse, error)
	// GetSubscription reads a subscription's detail.
	// GET /subscriptions/detail?subId=.
	GetSubscription(ctx context.Context, subID string) (*SubscriptionStatus, error)
	// GetCharges reads a subscription's charge ledger.
	// GET /subscriptions/charges?subId=&limit=&offset=.
	GetCharges(ctx context.Context, subID string, limit, offset int) (*ChargesResponse, error)
	// GetPendingChange reads a subscription's pending downgrade, or (nil, nil)
	// when there is none. GET /subscriptions/pending?subId=.
	GetPendingChange(ctx context.Context, subID string) (*PendingPlanChange, error)
}
